### PR TITLE
Added AWS Lambda Node.js instructions

### PR DESCRIPTION
**Pull request by michaelhyatt**
I added instructions similar to Cyrille's describing the Node.js Lambda setup.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
Add something like

IMPORTANT: due to a limitation in AWS Lambda Otel Layer for Node.js (see https://github.com/aws-observability/aws-otel-lambda/issues/118 ), it is required to create a trace in the calling invocation of the Node.js lambda functions in order to force the sampling strategy to `sampled:true` passing a `traceparent` header with `sampled=true`. 
Creating a trace in the calling invocation of the Lambda function can be achieved:
* If the caller is a web browser, instrumenting the caller with [Elastic RUM](https://www.elastic.co/guide/en/apm/get-started/current/rum.html) [enabling distributed tracing on the RUM agent](https://www.elastic.co/guide/en/apm/agent/rum-js/5.x/distributed-tracing-guide.html)
* If the caller is an application, instrumenting this application with an OpenTelemetry or Elastic APM agent ensuring distributed tracing is enabled

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 1]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+    traces:
+      receivers: [otlp]
+      exporters: [logging, otlp/elastic]
+    metrics:
+      receivers: [otlp]
+      exporters: [logging, otlp/elastic]
+----
+<1> Environment-specific configuration parameters can be conveniently passed in as environment variables: `ELASTIC_OTLP_ENDPOINT` and `ELASTIC_OTLP_TOKEN`
+
+* Configure the AWS Lambda Node.js function with:
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:356]:
```suggestion
2. Configure the AWS Lambda Node.js function:

```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 2]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+** https://docs.aws.amazon.com/lambda/latest/dg/API_Layer.html[Function
+layer]: The latest https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS
+Lambda layer for OpenTelemetry] (e.g. `arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-nodejs-ver-0-19-0:1`)
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:359]:
```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_Layer.html[Function
layer]: The latest https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS
Lambda layer for OpenTelemetry]. For example, `arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-nodejs-ver-0-19-0:1`)
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 3]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+** https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html[TracingConfig / Mode] set to `PassTrough`
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:360]:
```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html[TracingConfig / Mode] set to `PassTrough`
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 4]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+** https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionConfiguration.html[FunctionConfiguration / Timeout] set to more than 10 seconds to support the cold start of the Lambda JS Runtime
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:361]:
```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionConfiguration.html[FunctionConfiguration / Timeout] set to more than 10 seconds to support the cold start of the Lambda JS Runtime
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 5]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+** Export the environment variables:
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:362]:
```suggestion
* Export the environment variables:
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 6]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+*** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway (see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[here])
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:363]:
```suggestion
** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway. See https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[enable auto instrumentation for your lambda-function].
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 7]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+*** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:364]:
```suggestion
** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 8]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+*** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:365]:
```suggestion
** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 9]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+
+[[open-telemetry-aws-lambda-elastic-nodejs-terraform]]
+===== Instrumenting AWS Lambda Node.js functions with Terraform
+
+We recommend using an infrastructure as code solution like Terraform or Ansible to manage the configuration of your AWS Lambda functions.
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:370]:
```suggestion
To manage the configuration of your AWS Lambda functions, we recommend using an infrastructure as code solution like Terraform or Ansible.
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 10]
```diff
@@ -306,6 +306,73 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+
+Here is an example of AWS Lambda Node.js function managed with Terraform and the https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_function[AWS Provider / Lambda Functions]:
+
+* Sample Terraform code: https://github.com/michaelhyatt/terraform-aws-nodejs-api-worker-otel/tree/v0.19
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:374]:
```suggestion
* https://github.com/michaelhyatt/terraform-aws-nodejs-api-worker-otel/tree/v0.19[Sample Terraform code]
```

**Review by EamonnTP — COMMENTED** [review]
Hi @michaelhyatt This looks great! Thanks for adding it to the docs. Just a few minor suggestions.

File: docs/guide/opentelemetry-elastic.asciidoc (hunk 6 above)
Comment (cyrille-leclerc) [reply docs/guide/opentelemetry-elastic.asciidoc:365]: @EamonnTP why do you want to remove the 3rd level indentation of the bullet list?

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 11]
```diff
@@ -306,6 +306,75 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+
+To instrument AWS Lambda Node.js functions, see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS Distro for OpenTelemetry Lambda Support For JS].
+
+Noteworthy configuration elements:
+
+IMPORTANT: Due to a https://github.com/aws-observability/aws-otel-lambda/issues/118[limitation] in AWS Lambda Otel Layer for Node.js v0.19, you must create a trace in the calling invocation of the Node.js lambda functions. This is to force the sampling strategy to `sampled:true` passing a `traceparent` header with `sampled=true`. For example, with `-01` at the end of the value.
+
+To create a trace in the calling invocation of the Lambda function, use these or other methods that result in the `traceparent` header being added to the outgoing HTTP call:
+
+* If the caller is a web browser, instrument the caller using (https://www.elastic.co/guide/en/apm/get-started/current/rum.html)[Real User Monitoring (RUM)] by enabling https://www.elastic.co/guide/en/apm/agent/rum-js/5.x/distributed-tracing-guide.html[distributed tracing] on the RUM agent.
```
Comment (EamonnTP) [inline docs/guide/opentelemetry-elastic.asciidoc:322]:
```suggestion
* If the caller is a web browser, instrument the caller using <<rum,Real User Monitoring (RUM)>> by enabling {apm-rum-ref}/distributed-tracing-guide.html[distributed tracing] on the RUM agent.
```

File: docs/guide/opentelemetry-elastic.asciidoc (hunk 6 above)
Comment (EamonnTP) [reply docs/guide/opentelemetry-elastic.asciidoc:365]: Hi @cyrille-leclerc This is to fall inline with my other suggestions above; dropping the bullets point down a level. There are still a few other comments that need addressed. Thanks.

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 12]
```diff
@@ -306,6 +306,76 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
 * Note that the Terraform code to manage the HTTP API Gateway (https://github.com/cyrille-leclerc/my-serverless-shopping-cart/tree/main/utils/terraform/api-gateway-proxy[here]) is copied from the official OpenTelemetry Lambda sample https://github.com/open-telemetry/opentelemetry-lambda/tree/e72467a085a2a6e57af133032f85ac5b8bbbb8d1/utils[here]
 
+[[open-telemetry-aws-lambda-elastic-nodejs]]
+===== Instrumenting AWS Lambda Node.js functions
+
+NOTE: For a better startup time, we recommend using SDK-based instrumentation for manual instrumentation of the code rather than auto instrumentation.
+
+To instrument AWS Lambda Node.js functions, see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS Distro for OpenTelemetry Lambda Support For JS].
+
+Important configuration settings:
```
Comment (bmorelli25) [inline docs/guide/opentelemetry-elastic.asciidoc:316]:
```suggestion
```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 13]
```diff
@@ -306,6 +306,76 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+
+IMPORTANT: Due to a https://github.com/aws-observability/aws-otel-lambda/issues/118[limitation] in AWS Lambda Otel Layer for Node.js v0.19, you must create a trace in the calling invocation of the Node.js lambda functions. This is to force the sampling strategy to `sampled:true` passing a `traceparent` header with `sampled=true`. For example, with `-01` at the end of the value.
+
+To create a trace in the calling invocation of the Lambda function, use these or other methods that result in the `traceparent` header being added to the outgoing HTTP call:
+
+* If the caller is a web browser, instrument the caller using <<rum,Real User Monitoring (RUM)>> by enabling {apm-rum-ref}/distributed-tracing-guide.html[distributed tracing] on the RUM agent.
+* If the caller is an application, instrument the application with an OpenTelemetry or Elastic APM agent ensuring distributed tracing is enabled.
+
+1. The configuration of the OpenTelemetry Collector, with the definition of the Elastic Observability endpoint, can be added to the root directory of the Lambda binaries: `src/main/resources/opentelemetry-collector.yaml`.
++
```
Comment (bmorelli25) [inline docs/guide/opentelemetry-elastic.asciidoc:326]:
```suggestion
The configuration of the OpenTelemetry Collector, with the definition of the Elastic Observability endpoint, can be added to the root directory of the Lambda binaries: `src/main/resources/opentelemetry-collector.yaml`.

```

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 14]
```diff
@@ -306,6 +306,76 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+    traces:
+      receivers: [otlp]
+      exporters: [logging, otlp/elastic]
+    metrics:
+      receivers: [otlp]
+      exporters: [logging, otlp/elastic]
+----
+<1> Environment-specific configuration parameters can be conveniently passed in as environment variables: `ELASTIC_OTLP_ENDPOINT` and `ELASTIC_OTLP_TOKEN`
+
+2. Configure the AWS Lambda Node.js function:
```
Comment (bmorelli25) [inline docs/guide/opentelemetry-elastic.asciidoc:358]:
```suggestion
Configure the AWS Lambda Node.js function:
```

**Review by bmorelli25 — APPROVED** [review]
Thanks for writing this, @michaelhyatt! I committed @EamonnTP's suggestions but then had to add a few more formatting lines. Unfortunately, numbered lists don't render correctly this deep in nesting (it was an `a.` and `b.` instead).

**Comment by michaelhyatt** [comment]
@cyrille-leclerc @EamonnTP  there is a new release of OTel layer, v0.23 that changed few things. Is it worth updating the instructions?

**Comment by cyrille-leclerc** [comment]
@michaelhyatt I like the idea to be up to date if it doesn't require too much work

**Comment by michaelhyatt** [comment]
@cyrille-leclerc @EamonnTP I updated the instructions for v0.23 of the layer.

File: docs/guide/opentelemetry-elastic.asciidoc [hunk 15]
```diff
@@ -306,6 +306,70 @@ Here is an example of AWS Lambda Java function managed with Terraform and the ht
...
+** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway. See https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[enable auto instrumentation for your lambda-function].
+** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray
+** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration
+** `OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:55681/v1/traces"` this environment variable is required to be set until [ pull request](https://github.com/open-telemetry/opentelemetry-js/pull/2331) is merged and released.
+** `OTEL_TRACES_SAMPLER="AlwaysOn"` define the required sampler strategy if it is not sent from the caller. Note that `Always_on` can potentially create a very large amount of data, so in production set the correct sampling configuration, as per [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling).
```
Comment (bmorelli25) [inline docs/guide/opentelemetry-elastic.asciidoc:361]: Change from markdown to asciidoc links

```suggestion
** `OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:55681/v1/traces"` this environment variable is required to be set until https://github.com/open-telemetry/opentelemetry-js/pull/2331[PR #2331] is merged and released.
** `OTEL_TRACES_SAMPLER="AlwaysOn"` define the required sampler strategy if it is not sent from the caller. Note that `Always_on` can potentially create a very large amount of data, so in production set the correct sampling configuration, as per the https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling[specification].
```

**Review by bmorelli25 — APPROVED** [review]
Thanks! I converted the markdown links into asciidoc links

**Comment by bmorelli25** [comment]
@michaelhyatt, It looks like you used your personal account for the final commit, which is why the CLA is complaining. Would you mind signing the CLA with your personal email address?

**Comment by bmorelli25** [comment]
Verified with CLAchecker that all emails have signed the agreement

**Pull request merged**